lib/src/editor.dart

```diff
@@ -0,0 +1,499 @@
+  /// doc.parseAt(['b'], orElse: {"a": 42}); // YamlMap({"a": 42})
+  /// doc.parseAt(['b'], orElse: 42); // YamlScalar(42)
+  ///
+  /// final doc2 = YamlEditor('[0,1]');
+  /// doc2.parseAt([2]); // ArgumentError
+  /// doc2.parseAt(["2"]); // ArgumentError
+  /// ```
+  ///
+  /// [orElse] can also be used to determine if a path exists.
+  /// ```dart
+  /// doc.parseAt(my_path, orElse: #doesNotExist);
+  ///
+  /// // Or, if you know that your collection will not contain null,
+  /// doc.parseAt(my_path, orElse: null);
+  /// ```
+  @override
+  YamlNode parseAt(Iterable<Object> path, {Object orElse = #noArg}) {
+    try {
+      return _traverse(path);
+    } on ArgumentError {
+      if (orElse == #noArg) {
+        rethrow;
+      } else {
+        if (orElse is YamlNode) {
+          return orElse;
+        }
+
+        return yamlNodeFrom(orElse);
+      }
+    }
+  }
+
+  /// Sets [value] in the [path].
+  ///
+  /// Note that [assign] provides a different result as compared to a [remove] followed by an
+  /// [insertIntoList], because it preserves comments at the same level.
+  ///
+  /// Throws an [ArgumentError] if path is invalid, and throws the usual Dart errors otherwise (e.g.
+  /// [RangeError] if and index is negative or longer than list length).
+  ///
+  /// ```dart
+  /// final doc = YamlEditor('''
+  ///   - 0
+  ///   - 1 # comment
+  ///   - 2
+  /// ''');
+  /// doc.assign([1], 'test');
+  /// ```
+  ///
+  /// Expected Output:
+  /// ```yaml
+  ///   - 0
+  ///   - test # comment
+  ///   - 2
+  /// ```
+  ///
+  /// ```dart
+  /// final doc2 = YamlEditor('''
+  ///   - 0
+  ///   - 1 # comment
+  ///   - 2
+  /// ''');
+  /// doc2.remove([1]);
+  /// doc2.insertIntoList([], 1, 'test');
+  /// ```
+  ///
+  /// Expected Output:
+  /// ```yaml
+  ///   - 0
+  ///   - test
+  ///   - 2
+  /// ```
+  @override
+  void assign(Iterable<Object> path, Object value) {
+    if (path.isEmpty) {
+      final end = getContentSensitiveEnd(_contents);
+      final edit = SourceEdit(0, end, getFlowString(value));
+
+      _performEdit(edit, path, yamlNodeFrom(value));
+      return;
+    }
+
+    final collectionPath = path.take(path.length - 1);
+    final keyOrIndex = path.last;
+    final parentNode = parseAt(collectionPath);
+
+    var edit;
+    var expectedNode;
+
+    final valueNode = yamlNodeFrom(value);
+
+    if (parentNode is YamlList) {
+      edit = assignInList(_yaml, parentNode, keyOrIndex, value);
+      expectedNode =
+          updatedYamlList(parentNode, (nodes) => nodes[keyOrIndex] = valueNode);
+    } else if (parentNode is YamlMap) {
+      edit = assignInMap(_yaml, parentNode, keyOrIndex, value);
+      final keyNode = yamlNodeFrom(keyOrIndex);
+      expectedNode =
+          updatedYamlMap(parentNode, (nodes) => nodes[keyNode] = valueNode);
+    } else {
+      throw ArgumentError('Scalar $parentNode does not have key $keyOrIndex');
+    }
+
+    _performEdit(edit, collectionPath, expectedNode);
```

Comment:
nit: maybe, returning earlier is nicer:
```suggestion
    final valueNode = yamlNodeFrom(value);

    if (parentNode is YamlList) {
      _performEdit(
        assignInList(_yaml, parentNode, keyOrIndex, value),
        collectionPath,
        updatedYamlList(parentNode, (nodes) => nodes[keyOrIndex] = valueNode),
      );
      return;
    }

    if (parentNode is YamlMap) {
      final keyNode = yamlNodeFrom(keyOrIndex);
      _performEdit(
        assignInMap(_yaml, parentNode, keyOrIndex, value),
        collectionPath,
        updatedYamlMap(parentNode, (nodes) => nodes[keyNode] = valueNode),
      );
      return;
    }

    throw ArgumentError('Scalar $parentNode does not have key $keyOrIndex');
```